app/src/main/java/com/pileproject/drive/execution/ExecutionCondition.java

```diff
@@ -17,48 +17,186 @@
-     *
-     * @param result
+     * Check the program count is over than the program size or not.
+     * @return finished (true) or not (false)
+     */
+    public boolean hasProgramFinished() {
+        return mProgramCount < mBlocks.size();
+    }
+
+    /**
+     * Get the current block.
+     * @return
+     */
+    public BlockBase getCurrentBlock() {
+        return mBlocks.get(mProgramCount);
+    }
+
+    /**
+     * Increment the program count.
+     */
+    public void incrementProgramCount() {
+        mProgramCount++;
+    }
+
+    /**
+     * Decrement the program count.
+     */
+    public void decrementProgramCount() {
+        mProgramCount--;
+    }
+
+    /**
+     * Set the program count.
+     * @param pc
+     */
+    public void setProgramCount(int pc) {
+        if (pc < 0 || pc >= mBlocks.size())
+            throw new IndexOutOfBoundsException("The program count is invalid.");
+    }
+
+    /**
+     * Get the current program count.
+     * @return program count
+     */
+    public int getProgramCount() {
+        return mProgramCount;
+    }
+
+    /**
+     * Push the pair of (the index of the selection block, the result: true or false) to ifStack
+     * @param result the result of a selection command
      */
     public void pushSelectionResult(boolean result) {
-        Map<String, Integer> map = new HashMap<>();
-        Context context = DriveApplication.getContext();
-        map.put(context.getString(R.string.key_execution_index), programCount);    // current index
-        map.put(context.getString(R.string.key_execution_result), result ? TRUE : FALSE);
-        ifStack.push(map);
+        IfStatus status = new IfStatus(mProgramCount, result);
+        mIfStack.push(status);
+    }
+
+    /**
+     * Pop and throw away the latest selection result.
+     */
+    public void popSelectionResult() {
+        mIfStack.pop();
+    }
+
+    /**
+     * Push the index of the beginning block of the current loop to a stack.
+     * @param index the index of the beginning block
+     */
+    public void pushBeginningOfLoop(int index) {
+        mWhileStack.push(index);
+    }
+
+    /**
+     * Check the current block should be through or not.
+     *
+     * @return through (true) or not (false)
+     */
+    public boolean isThrough() {
+        if (mIfStack.isEmpty()) return false;
+
+        BlockBase block = mBlocks.get(mProgramCount);
+        if (block instanceof SelectionEndBlock) return false;
+
+        IfStatus ifStatus = mIfStack.peek();
+        BlockBase nearestSelectionBlock = mBlocks.get(ifStatus.index);
+
+        int middleIf = (nearestSelectionBlock.getRight() + nearestSelectionBlock.getLeft()) / 2;
+        int middle = (block.getRight() + block.getLeft()) / 2;
+
+        // go through this block if (true && left) || (!false && right)
+        return (ifStatus.result && middle <= middleIf) || (!ifStatus.result && middleIf < middle);
```

Comment:
The best way I came up with is to make a new class that holds graphics-specific-if logic, which might implement an interface common with other strategies.
Or the logic could be in <code>Block</code> class?

I think in future we should separate geometric information from <code>BlockBase</code> class because we are going to implement text-based programming interface in our app.
We can use one <code>BlockBase</code> instance as an atomic operation, which does not need geometric information. When users write some code in our app, then we should parse it, convert it into list of <code>BlockBase</code> instances then we can reuse our code base we already have (though we might need more pretty class name).

So the geometric information should leave from the block class and a new interface, say <code>VisualBlock</code> interface or whatever, should be responsible for the information.
Then the if-logic should be in the interface.

But if we could compile a program and could know next indexes of if-blocks, this conversation would be useless.